source-firestore/main.go

```diff
@@ -0,0 +1,263 @@
+	var cfg config
+	if err := args.ConfigFile.Parse(&cfg); err != nil {
+		return err
+	}
+
+	ctx := context.Background()
+	sa := option.WithCredentialsJSON([]byte(cfg.CredentialsJSON))
+	app, err := firebase.NewApp(ctx, nil, sa)
+	if err != nil {
+		return err
+	}
+
+	client, err := app.Firestore(ctx)
+	if err != nil {
+		return err
+	}
+	defer client.Close()
+
+	var catalog = new(airbyte.Catalog)
+	var collections = client.Collections(ctx)
+	for {
+		var collection, err = collections.Next()
+		if err == iterator.Done {
+			break
+		} else if err != nil {
+			return err
+		}
+
+		var query = collection.Query.Limit(DISCOVER_DOC_LIMIT)
+		var docs = query.Documents(ctx)
+
+		var docsCh = make(chan schema_inference.Document)
+		var logEntry = log.WithField("collection", collection.ID)
+
+		var eg = new(errgroup.Group)
+		var schema schema_inference.Schema
+
+		eg.Go(func() error {
+			schema, err = schema_inference.Run(ctx, logEntry, docsCh)
+			if err != nil {
+				return fmt.Errorf("schema inference: %w", err)
+			}
+			return nil
+		})
+
+		for {
+			var doc, err = docs.Next()
+			if err == iterator.Done {
+				break
+			} else if err != nil {
+				return err
+			}
+			var data = doc.Data()
+			data[firestore.DocumentID] = doc.Ref.ID
+			docJson, err := json.Marshal(data)
+			docsCh <- docJson
+		}
+		close(docsCh)
+
+		if err := eg.Wait(); err != nil {
+			return err
+		}
+
+		catalog.Streams = append(catalog.Streams, airbyte.Stream{
+			Name:                    collection.ID,
+			JSONSchema:              schema,
+			SupportedSyncModes:      []airbyte.SyncMode{airbyte.SyncModeIncremental},
+			SourceDefinedCursor:     true,
+			SourceDefinedPrimaryKey: [][]string{{firestore.DocumentID}},
+		})
+	}
+
+	var encoder = airbyte.NewStdoutEncoder()
+	return encoder.Encode(airbyte.Message{
+		Type:    airbyte.MessageTypeCatalog,
+		Catalog: catalog,
+	})
+}
+
+func doRead(args airbyte.ReadCmd) error {
+	var cfg config
+	var state State
+	var catalog airbyte.ConfiguredCatalog
+
+	if err := args.ConfigFile.Parse(&cfg); err != nil {
+		return err
+	} else if err := args.CatalogFile.Parse(&catalog); err != nil {
+		return err
+	} else if args.StateFile != "" {
+		if err := args.StateFile.Parse(&state); err != nil {
+			return err
+		}
+	}
+
+	if state.Snapshots == nil {
+		state.Snapshots = map[string]string{}
+	}
+
+	ctx := context.Background()
+	sa := option.WithCredentialsJSON([]byte(cfg.CredentialsJSON))
+	app, err := firebase.NewApp(ctx, nil, sa)
+	if err != nil {
+		return err
+	}
+
+	client, err := app.Firestore(ctx)
+	if err != nil {
+		return err
+	}
+	defer client.Close()
+
+	var enc = airbyte.NewStdoutEncoder()
+	for {
+		var now = time.Now()
+		for _, stream := range catalog.Streams {
+			var streamName = stream.Stream.Name
+			var streamState = state.Snapshots[streamName]
+
+			var collection = client.Collection(streamName)
+			// FIXME: DocumentID is not ordered, so the user must have a field for us to order the documents on
```

Comment:
This is the main challenge: Firestore's default document identifier is not ordered, from their [docs](https://firebase.google.com/docs/firestore/manage-data/add-data):

> Important: Unlike "push IDs" in the Firebase Realtime Database, Cloud Firestore auto-generated IDs do not provide any automatic ordering. If you want to be able to order your documents by creation date, you should store a timestamp as a field in the documents.

So the user needs to provide us with a column which we can use for chronological ordering of documents.

Given the Airbyte protocol, I'm not sure how can we allow users to configure a cursor for each stream? I see [these two fields](https://github.com/estuary/flow/blob/master/go/protocols/airbyte/catalog.go#L21-L22) in the protocol definition, but I'm not sure how that's mapped from `resource_spec`, or is it at all?

If we can't make this configurable per stream, it needs to be universal across all collections which might be more difficult for users.